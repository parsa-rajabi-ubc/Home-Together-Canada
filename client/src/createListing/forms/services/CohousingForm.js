/**
 * @Author:     Jeff Hatton
 * @Created:    2021.02.10
 *
 * @Description: Cohousing listing form
 *
 */

import React, {useEffect, useState} from 'react';
import PropTypes from "prop-types";
import {COHOUSING_TEXT as TEXT} from "./constants/ServiceListingText";
import TextArea from "../../../common/forms/TextArea";
import LabelAsterisk from "../../../common/forms/LabelAsterisk";
import LargeTextArea from "../../../common/forms/LargeTextArea";
import SubmitButton from "../../../common/forms/SubmitButton";
import {SHORT_DESC_CHAR_COUNT} from "../../../common/constants/listingsConstants";
import {
    checkIfErrorsExistInMapping,
    validateInput,
    validatePostalCode
} from "../../../registration/registrationUtils";
import {validatePositiveNumber} from "../../../common/utils/generalUtils";

const CohousingForm = (props) => {
    const { onSubmit } = props;

    const [title, setTitle] = useState(undefined);
    const [contactName, setContactName] = useState(undefined);
    const [shortDescription, setShortDescription] = useState(undefined);
    const [fullDescription, setFullDescription] = useState(undefined);
    const [unitsForSale, setUnitsForSale] = useState(undefined);
    const [unitsForRent, setUnitsForRent] = useState(undefined);

    const [titleError, setTitleError] = useState(undefined);
    const [contactNameError, setContactNameError] = useState(undefined);
    const [shortDescriptionError, setShortDescriptionError] = useState(undefined);
    const [fullDescriptionError, setFullDescriptionError] = useState(undefined);
    const [unitsForSaleError, setUnitsForSaleError] = useState(undefined);
    const [unitsForRentError, setUnitsForRentError] = useState(undefined);

    const shortDescCharCount = {SHORT_DESC_CHAR_COUNT};

    useEffect(() => {
        title !== undefined && validateInput(title, setTitleError);
    }, [title]);
    useEffect(() => {
        contactName !== undefined && validatePostalCode(contactName, setContactNameError);
    }, [contactName]);
    useEffect(() => {
        shortDescription !== undefined && validateInput(shortDescription, setShortDescriptionError);
    }, [shortDescription]);
    useEffect(() => {
        fullDescription !== undefined && validateInput(fullDescription, setFullDescriptionError);
    }, [fullDescription]);
    useEffect(() => {
        unitsForSale !== undefined && validatePositiveNumber(unitsForSale, setUnitsForSaleError);
    }, [unitsForSale]);
    useEffect(() => {
        unitsForRent !== undefined && validatePositiveNumber(unitsForRent, setUnitsForRentError);
    }, [unitsForRent]);

    const isFormValid = () => {

        const errors = {
            title: false,
            contactName: false,
            shortDes: false,
            fullDes: false,
            unitsForSale: false,
            unitsForRent: false,
        }

        errors.title = validateInput(title, setTitleError);
        errors.contactName = validateInput(contactName, setContactNameError);
        errors.shortDes = validateInput(shortDescription, setShortDescriptionError);
        errors.fullDes = validateInput(fullDescription, setFullDescriptionError);
        errors.unitsForSale = validatePositiveNumber(unitsForSale, setUnitsForSaleError);
        errors.unitsForRent = validatePositiveNumber(unitsForRent, setUnitsForRentError);

        return !(checkIfErrorsExistInMapping(errors));
    }

    //function for input checks on submit
    function onCreateListing() {

        if (!isFormValid()) {
            console.log("Form is invalid")
            // form is invalid
        }
        // else form is valid and proceed to make request
        else {
            onSubmit();
        }
    }

    return (
        <div>
            <div className="col-start-1 col-end-7 py-5 px-5 m-6 bg-white shadow-lg rounded-xl">
                <div className="grid grid-cols-2 gap-6">
                    <div className="col-span-3 sm:col-span-2">

                        <h1 className={"page-title mb-5"}> {TEXT.form_title} </h1>

                        <TextArea
                            className={`${titleError && "border-red-500"} input`}
                            label={TEXT.title}
                            labelClassName={"label"}
                            required={true}
                            onChange={(e) => setTitle(e.target.value)}
                        />

                        <div className={"grid grid-cols-9 gap-x-6"}>

                            <section className={"col-start-1 col-end-5"}>

                                <TextArea
                                    className={`${shortDescriptionError && "border-red-500"} input`}
                                    label={TEXT.short_des + " (" + SHORT_DESC_CHAR_COUNT + " Characters)"}
                                    labelClassName={"label"}
                                    required={true}
                                    onChange={(e) => setShortDescription(e.target.value)}
                                    charLimit={shortDescCharCount}
                                />
                            </section>
                            <section className={"col-start-6 col-end-9"}>
                                <TextArea
                                    className={`${contactNameError && "border-red-500"} input`}
                                    label={TEXT.contact_name}
                                    labelClassName={"label"}
                                    required={true}
                                    onChange={(e) => setContactName(e.target.value)}
                                />
                            </section>

                            <section className={"col-start-1 col-end-5"}>
                                <LabelAsterisk
                                    label={TEXT.units_for_sale}
                                    className={"label"}
                                />
                                <input
                                    className={`${unitsForSaleError && "border-red-500"} input`}
                                    type="number"
                                    min="0"
                                    step="1"
                                    onChange={(e) => setUnitsForSale(e.target.value)}
                                />

                            </section>
                            <section className={"col-start-1 col-end-5"}>
                                <LabelAsterisk
                                    label={TEXT.units_for_rent}
                                    className={"label"}
                                />
                                <input
                                    className={`${unitsForRentError && "border-red-500"} input`}
                                    type="number"
                                    min="0"
                                    step="1"
                                    onChange={(e) => setUnitsForRent(e.target.value)}
                                />
                            </section>
                        </div>

                        <LargeTextArea
                            className={`${fullDescriptionError && "border-red-500"} input`}
                            rows={"6"}
                            label={TEXT.full_des}
                            labelClassName={"label"}
                            required={true}
                            onChange={(e) => setFullDescription(e.target.value)}
                        />
                    </div>
                </div>
            </div>
            <SubmitButton className={"btn btn-green form-btn w-1/2"} onClick={onCreateListing}
                          onSubmit={onCreateListing}/>
        </div>
    )

}
CohousingForm.propTypes = {
    onSubmit: PropTypes.func.isRequired
}

export default CohousingForm;