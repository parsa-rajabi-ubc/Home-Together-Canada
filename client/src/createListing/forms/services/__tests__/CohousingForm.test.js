/**
 * @Author:     Jeff Hatton
 * @Created:    2021.2.11
 *
 * @Description: Cohousing Form Test
 *
 */

import React from 'react';
import renderer from  'react-test-renderer'
import CohousingForm from "../CohousingForm";

describe('CohousingForm', () => {
    describe('Snapshot test', () => {
        it("should render correctly regardless of properties", () => {
            // given
            const props = {
                onSubmit: jest.fn()
            };
            //when
            const component = renderer.create(<CohousingForm {...props}/>).toJSON();
            //then
            expect(component).toMatchSnapshot();
        });
    })
})