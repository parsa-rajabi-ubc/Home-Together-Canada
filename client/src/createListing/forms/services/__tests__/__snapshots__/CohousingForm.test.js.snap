// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CohousingForm Snapshot test should render correctly regardless of properties 1`] = `
<div>
  <div
    className="col-start-1 col-end-7 py-5 px-5 m-6 bg-white shadow-lg rounded-xl"
  >
    <div
      className="grid grid-cols-2 gap-6"
    >
      <div
        className="col-span-3 sm:col-span-2"
      >
        <h1
          className="page-title mb-5"
        >
           
          Cohousing, Co-ops, Intergenerational, Planned Neighbourhoods
           
        </h1>
        <section>
          <label
            className="label"
          >
            Descriptive Title
          </label>
          <svg
            className="inline align-top text-red-600"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
              fill="currentColor"
            />
          </svg>
          <input
            className="undefined input"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </section>
        <div
          className="grid grid-cols-9 gap-x-6"
        >
          <section
            className="col-start-1 col-end-5"
          >
            <section>
              <label
                className="label"
              >
                Short Description (50 Characters)
              </label>
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
              <input
                className="undefined input"
                disabled={false}
                maxLength={
                  Object {
                    "SHORT_DESC_CHAR_COUNT": "50",
                  }
                }
                onChange={[Function]}
                type="text"
              />
            </section>
          </section>
          <section
            className="col-start-6 col-end-9"
          >
            <section>
              <label
                className="label"
              >
                Contact Name
              </label>
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
              <input
                className="undefined input"
                disabled={false}
                onChange={[Function]}
                type="text"
              />
            </section>
          </section>
          <section
            className="col-start-1 col-end-5"
          >
            <label
              className="label"
            >
              Number of units for sale
               
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
            </label>
            <input
              className="undefined input"
              min="0"
              onChange={[Function]}
              step="1"
              type="number"
            />
          </section>
          <section
            className="col-start-1 col-end-5"
          >
            <label
              className="label"
            >
              Number of units for rent
               
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
            </label>
            <input
              className="undefined input"
              min="0"
              onChange={[Function]}
              step="1"
              type="number"
            />
          </section>
        </div>
        <div>
          <label
            className="label"
          >
            Full Description
          </label>
          <svg
            className="inline align-top text-red-600"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
              fill="currentColor"
            />
          </svg>
          <textarea
            className="undefined undefined input"
            cols="50"
            onChange={[Function]}
            rows="6"
          />
        </div>
      </div>
    </div>
  </div>
  <label>
    <input
      className="btn btn-green form-btn w-1/2"
      onClick={[Function]}
      type="submit"
    />
  </label>
</div>
`;
